kitsune/inproduct/admin.py

```diff
@@ -1,13 +1,84 @@
+from django.conf import settings
 from django.contrib import admin
+from django.utils.html import format_html
+from django.utils.translation import gettext_lazy as _
 
 from kitsune.inproduct.models import Redirect
+from kitsune.wiki.models import Revision
+
+
+class KBLinkFilter(admin.SimpleListFilter):
+    title = _("KB article")
+    parameter_name = "kb_article"
+
+    def lookups(self, request, model_admin):
+        return (
+            ("with_kb", _("Has KB Article")),
+            ("without_kb", _("No KB Article")),
+        )
+
+    def queryset(self, request, queryset):
+        if self.value() == "with_kb":
+            # Filter for objects with a KB link
+            return queryset.filter(id__in=[obj.id for obj in queryset if self.has_kb_link(obj)])
+        if self.value() == "without_kb":
+            # Filter for objects without a KB link
+            return queryset.exclude(id__in=[obj.id for obj in queryset if self.has_kb_link(obj)])
+
+    def has_kb_link(self, obj):
+        if "kb/" in obj.target:
+            try:
+                doc_slug = obj.target.split("/")[-1]
+                if obj.locale:
+                    locale = obj.locale
+                else:
+                    locale = settings.WIKI_DEFAULT_LANGUAGE
+                rev = Revision.objects.filter(
+                    document__slug=doc_slug, document__locale=locale
+                ).exists()
+                return rev
+            except Revision.DoesNotExist:
+                pass
+        return False
```

Comment:
I really like the use of a `SimpleListFilter`! It'd be nice, however, to avoid the hundreds of `Revision` queries this would entail, one for each redirect. I think this can be done in a single query. I tried the following, and it worked:

```suggestion
    def queryset(self, request, queryset):
        qs = queryset.annotate(
            kb_slug=Case(
                When(target__startswith="kb/", then=Replace("target", Value("kb/"))),
                When(target__startswith="/kb/", then=Replace("target", Value("/kb/"))),
                default=None,
            ),
            kb_locale=Case(
                When(locale="", then=Value(settings.WIKI_DEFAULT_LANGUAGE)),
                default=F("locale"),
            ),
        ).filter(kb_slug__isnull=False)
        doc_exists = Exists(
            Document.objects.filter(slug=OuterRef("kb_slug"), locale=OuterRef("kb_locale"))
        )
        if self.value() == "with_kb":
            # Filter for objects with a KB link
            return qs.filter(doc_exists)
        if self.value() == "without_kb":
            # Filter for KB objects without a KB link
            return qs.exclude(doc_exists)
```

Reply:
Agreed.